### PR TITLE
[Feat] Add Muon Support Phase 1 (dp without zero)

### PR DESCRIPTION
**Pull request by 0xWJ**
1. Fix tracing error for pytorch 2.9+
2. Fix state dicts related logic for Muon.
3. Add MASTER_PORT (also used as PYTEST_RUN_ID) env for tests, so we can run multiple pytest in the same machine.

### CODE REVIEW

**Comment by zyeric** [comment]
> 1. Fix tracing error for pytorch 3.9+
> 2. Fix state dicts related logic for Muon.
> 3. Add MASTER_PORT (also used as PYTEST_RUN_ID) env for tests, so we can run multiple pytest in the same machine.

tracing error for 2.9+?

File: nnscaler/runtime/utils.py [hunk 1]
```diff
@@ -104,3 +105,114 @@ def microbatches(samples: List[Any], cycle: bool = False) -> MicroBatchDataLoade
...
+            result_idx[-1].append(i)
+            current_sum += x
+
+    return result, result_idx
+
+
+def _split_array_min_max_out_of_order(nums: list[int], g: int) -> tuple[list[list[int]], list[list[int]]]:
+    """
+    Split the array nums into g subarrays (order of elements can be changed)
+    This problem (multi-way number partitioning) is NP-hard. We use a greedy approximation algorithm here.
```
Comment (zyeric) [inline nnscaler/runtime/utils.py:192]: how good is this approximation? (the coefficient with respect to the optimal value)

File: nnscaler/runtime/utils.py [hunk 2]
```diff
@@ -104,3 +105,114 @@ def microbatches(samples: List[Any], cycle: bool = False) -> MicroBatchDataLoade
         MicroBatchDataLoader: a micro-batch data loader.
     """
     return MicroBatchDataLoader(samples, cycle=cycle)
+
+
+def split_array_min_max(nums: list[int], g: int, *, keep_order: bool = True) -> tuple[list[list[int]], list[list[int]]]:
```
Comment (zyeric) [inline nnscaler/runtime/utils.py:110]: when will this function be used?

Comment (0xWJ) [reply nnscaler/runtime/utils.py:110]: Oh, This function will be used to support zero. My plan is to partition continous buffer in param boundary.

File: nnscaler/runtime/utils.py (hunk 1 above)
Comment (0xWJ) [reply nnscaler/runtime/utils.py:192]: From document, it is good enough.  the overhead is about 4% (padding) for a real MoE example.

**Comment by 0xWJ** [comment]
> > 1. Fix tracing error for pytorch 3.9+
> > 2. Fix state dicts related logic for Muon.
> > 3. Add MASTER_PORT (also used as PYTEST_RUN_ID) env for tests, so we can run multiple pytest in the same machine.
> 
> tracing error for 2.9+?

See concrete_tracer.py.  Copied from you code.

File: nnscaler/runtime/utils.py (hunk 1 above)
Comment (zyeric) [reply nnscaler/runtime/utils.py:192]: I mean since it is a well studied np-hard problem, each approximation algorithm has a bound in theory

**Comment by zyeric** [comment]
> > > 1. Fix tracing error for pytorch 3.9+
> > > 2. Fix state dicts related logic for Muon.
> > > 3. Add MASTER_PORT (also used as PYTEST_RUN_ID) env for tests, so we can run multiple pytest in the same machine.
> > 
> > 
> > tracing error for 2.9+?
> 
> See concrete_tracer.py. Copied from you code.

I think it is a typo, should be 2.9+. Thanks

**Comment by 0xWJ** [comment]
> > > > 1. Fix tracing error for pytorch 3.9+
> > > > 2. Fix state dicts related logic for Muon.
> > > > 3. Add MASTER_PORT (also used as PYTEST_RUN_ID) env for tests, so we can run multiple pytest in the same machine.
> > > 
> > > 
> > > tracing error for 2.9+?
> > 
> > 
> > See concrete_tracer.py. Copied from you code.
> 
> I think it is a typo, should be 2.9+. Thanks
Haha, fixed.

File: nnscaler/runtime/utils.py (hunk 1 above)
Comment (0xWJ) [reply nnscaler/runtime/utils.py:192]: https://en.wikipedia.org/wiki/Greedy_number_partitioning

```
An improved greedy algorithm is called [LPT scheduling]. 
It processes the inputs by descending order of value, from large to small. 

Since it needs to pre-order the inputs, it can be used only as an [offline algorithm]. 
It guarantees that the largest sum is at most (4k-1)/3k  times the optimal (minimum) largest sum, 
and the smallest sum is at least  (3k-1)/(4k-2) times the optimal (maximum) smallest sum. 

See [LPT scheduling](https://en.wikipedia.org/wiki/LPT_scheduling) for more details.
```

**Review by lynex — COMMENTED** [review]
Please update requirements-dev.txt by adding following packages as used by modified test_*.py in this PR

more_itertools
dill
psutil

**Comment by 0xWJ** [comment]
> Please update requirements-dev.txt by adding following packages as used by modified test_*.py in this PR
> 
> more_itertools dill psutil

they are already in requirements.txt

**Review by lynex — COMMENTED** [review]
> > Please update requirements-dev.txt by adding following packages as used by modified test_*.py in this PR
> > 
> > more_itertools dill psutil
> 
> they are already in requirements.txt 

Understood. The two requirements files are now dedicated to runners and developers.

File: nnscaler/runtime/utils.py (hunk 1 above)
Comment (zyeric) [reply nnscaler/runtime/utils.py:192]: can you add the link and explanation to the comment, thanks.

Comment (0xWJ) [reply nnscaler/runtime/utils.py:192]: added

**Review by zyeric — APPROVED** [review]
